### ISSUE TITLE
fuse_main_real: Control is returned before Flush called

### ISSUE THREAD

**Issue by kalemas** [billziss-gh/winfsp#365]
## Bug Report

Hello, we have an issue with `rclone mount :local:f: p:` when writing a file: control is returned to user before `Flush` so no error is reported to user.

Thread on rclone forum: https://forum.rclone.org/t/rclone-mount-synchonous-write-transaction/23771/22

Investigating i found that Flush is called on Close handler if i start with fuse_main_real (like rclone do) and on Cleanup if i setup with FspFileSystemCreate (actually with winfsPy). It probably Flush should be called on Close handler to conform fuse compatibility but control should be blocked before Close?
 
[https://forum.rclone.org/uploads/short-url/mCUVAsDsZGDETLdlPYMl99MXHYP.log](rclone.log)

### How to Reproduce

Mount filesystem with rclone mount at write restricted directory, write a file with `echo 1 > p:\test.txt`, no error is returned.

### Behaviors

Expected permission denied error.

### Environment

- OS version and build: 20H2 x64
- WinFsp version and build: 2020.2

**Comment by billziss-gh** [comment]
Sorry for the long delay in responding.

> Investigating i found that Flush is called on Close handler if i start with fuse_main_real (like rclone do) and on Cleanup if i setup with FspFileSystemCreate (actually with winfsPy). It probably Flush should be called on Close handler to conform fuse compatibility but control should be blocked before Close?

Unfortunately you have found an area where the Linux and Windows FUSE implementations diverge.

- On Linux the `flush` happens when the file descriptor is closed ([`filp_close`](https://github.com/torvalds/linux/blob/v4.20/fs/open.c#L1139-L1140)). This allows a file system to report a last minute error. (But please note that the situation is more complicated when doing memory mapped I/O.)

- On Windows there is no direct equivalent to the `flush` operation. The Windows file system has a `Cleanup` operation, which in general has other responsibilities (e.g. dealing with file deletion) and a `Close` operation (which is analogous to the Linux `release`). Crucially the return code of both `Cleanup` and `Close` operations is ignored, so any error from `flush` will always be ignored.

In general I recommend that FUSE file systems do not rely on `flush`. It only works well on Linux and even on Linux there are cases where `flush` may not be called at an appropriate time (e.g. in this pattern: `open`, `mmap`, `close`, perform I/O via memory access, `munmap`).

> Mount filesystem with rclone mount at write restricted directory, write a file with echo 1 > p:\test.txt, no error is returned.

I would have to understand more about your scenario as well as how rclone handles this particular case. For example, is kernel-level file caching enabled (WinFsp-FUSE parameter `-oFileInfoTimeout=-1`)?

**Comment by kalemas** [comment]
With that option i have errors even on writing into writeable folder
[rclone-fileinfotimeout-1.log](https://github.com/billziss-gh/winfsp/files/6518352/rclone-fileinfotimeout-1.log)
it may be rclone issue, but control is still returned before flushing (inspected with breakpoint) and file does appear (only on mirrored filesystem that doesn't any matter).

It all sound not optimistic.

I m not realize what it means to do not _rely on `flush`_. Should windows implementations restrict asynchronous writing and report errors on `Write`?

I m successfully handle error in Cleanup when implementing over https://github.com/Scille/winfspy/blob/master/src/winfspy/memfs.py so i do not realize where `Cleanup` error code is ignored and why it should be ignored by driver.

**Comment by billziss-gh** [comment]
> but control is still returned before flushing

This is by design when you have kernel caching enabled. The following scenario is typical:

- App issues `WriteFile` requests. They are all buffered and satisfied by kernel cache (i.e. they are a simple `memcpy` into the kernel buffers).
- App closes file with `CloseHandle`. However the OS keeps the file open under the hood.
- A few moments later the "lazy writer" will wake up and WRITE buffered data to the file system. If these WRITEs fail then the lazy writer will try again. If the problem persists you will eventually get the infamous "Delayed Write Failed" error (see [here](https://docs.microsoft.com/en-US/windows-server/storage/file-server/troubleshoot/troubleshoot-event-id-50-error)).

The best way for an application to guarantee that this problem cannot happen is to issue `FlushFileBuffers` prior to closing the file (this will appear to a FUSE file system as `fsync` not `flush`).

This problem is not unique to Windows. Note this text from the [`close(2)` Linux man page](https://linux.die.net/man/2/close):

> A successful close does not guarantee that the data has been successfully saved to disk, as the kernel defers writes. It is not common for a file system to flush the buffers when the stream is closed. If you need to be sure that the data is physically stored use fsync(2). (It will depend on the disk hardware at this point.)

> I m not realize what it means to do not rely on flush

It means that when I design my own file systems at least, I never rely on `flush` as being a reliable last minute operation before `release`. This is because `flush` may or may not be called after `write` even on Linux (as explained in the case of memory mapped files it is possible to have `write` followed by `release` without `flush`).

> i do not realize where Cleanup error code is ignored and why it should be ignored by driver.

It is not ignored by the driver, it is ignored by Windows. Windows is not Linux and just because Linux issues `flush` before `release` on `close(2)` does not mean that Windows must do so as well.

**Comment by kalemas** [comment]
> This is by design when you have kernel caching enabled.
i do not see any option in winfsp params passed from rclone related to kernel caching: `-o attr_timeout=1 -o debug -o uid=-1 -o gid=-1 --FileSystemName=rclone -o volname=local   ? F`. Is kernel caching enabled by default, and is there way to disable it to test if the issue persists?

I discovered some issues related to rclone and already reported them.

**Comment by billziss-gh** [comment]
> i do not see any option in winfsp params passed from rclone related to kernel caching

You are correct: The `-o attr_timeout=1` results in `FileInfoTimeout=1000` which enables kernel metadata caching but NOT kernel data caching. This means that the lazy writer will NOT be used (it is only used when kernel data caching is enabled).

**Comment by ncw** [comment]
Hi Bill!

> * On Linux the `flush` happens when the file descriptor is closed ([`filp_close`](https://github.com/torvalds/linux/blob/v4.20/fs/open.c#L1139-L1140)). This allows a file system to report a last minute error. (But please note that the situation is more complicated when doing memory mapped I/O.)
>
> * On Windows there is no direct equivalent to the `flush` operation. The Windows file system has a `Cleanup` operation, which in general has other responsibilities (e.g. dealing with file deletion) and a `Close` operation (which is analogous to the Linux `release`). Crucially the return code of both `Cleanup` and `Close` operations is ignored, so any error from `flush` will always be ignored.

What appears to be happening is that rclone is returning an error to cgofuse/WinFSP from the flush call (note the `errc=-5` below)

```
2021/05/22 15:52:43 DEBUG : /test.txt: Flush: fh=0x0
2021/05/22 15:52:43 DEBUG : Local file system at //?/F:/: File to upload is small (6 bytes), uploading instead of streaming
2021/05/22 15:52:43 ERROR : test.txt: Failed to copy: The system cannot find the file specified.
2021/05/22 15:52:43 ERROR : test.txt: WriteFileHandle.New Rcat failed: The system cannot find the file specified.
2021/05/22 15:52:43 ERROR : test.txt: WriteFileHandle.Flush error: The system cannot find the file specified.
2021/05/22 15:52:43 ERROR : IO error: The system cannot find the file specified.
2021/05/22 15:52:43 DEBUG : /test.txt: >Flush: errc=-5
```

But this is never making it back to the application.

Is this expected behaviour on Windows? On Linux the return value from the `Flush` call will always be reported back to the application whereas the return value of the `Release` call will never be reported and I'd argue that failing to report errors on `Flush` back to the application is a bug as it differs from the the expectations of FUSE file systems.

> In general I recommend that FUSE file systems do not rely on `flush`. It only works well on Linux and even on Linux there are cases where `flush` may not be called at an appropriate time (e.g. in this pattern: `open`, `mmap`, `close`, perform I/O via memory access, `munmap`).

Rclone (in this case without a local disk buffer) can only return errors after it has finished writing (or attempted to write) the file. It can only do that after the last close which under Linux at least translates to the `Flush` call. Before `Flush` is called we don't know if any more writes will be coming.

**Comment by billziss-gh** [comment]
> What appears to be happening is that rclone is returning an error to cgofuse/WinFSP from the flush call (note the errc=-5 below). But this is never making it back to the application.

Yes, on Windows the `Flush` error will be ignored.

> I'd argue that failing to report errors on Flush back to the application is a bug as it differs from the the expectations of FUSE file systems.

Perhaps yes, but I do not believe that there is a way to fix the problem (would love to be proven wrong here).

> Rclone (in this case without a local disk buffer) can only return errors after it has finished writing (or attempted to write) the file. It can only do that after the last close which under Linux at least translates to the Flush call. Before Flush is called we don't know if any more writes will be coming.

The problem with `flush` is that it is not guaranteed to arrive before a `release`, even when `write` is performed. Consider this scenario (untested, but encapsulates behavior I have seen over the years in Linux and macOS).

```C
fd = open(...);
p = mmap(..., fd, ...);
e = close(fd); // Linux will send flush to the file system, but it will succeed since nothing has been written
assert(0 == e);
*p = 42;
munmap(p, ...); // Linux will eventually send release to the file system, but without additional flush
```

**Comment by kalemas** [comment]
@billziss-gh what do you think about following
```diff
diff --git a/src/dll/fuse/fuse_intf.c b/src/dll/fuse/fuse_intf.c
index b13c8541..630da8ba 100644
--- a/src/dll/fuse/fuse_intf.c
+++ b/src/dll/fuse/fuse_intf.c
@@ -1191,11 +1194,23 @@ static VOID fsp_fuse_intf_Cleanup(FSP_FILE_SYSTEM *FileSystem,
             if (0 != f->ops.unlink)
                 f->ops.unlink(filedesc->PosixPath);
         }
+    if (Flags & FspCleanupSetLastWriteTime && !filedesc->IsDirectory)
+    {
+        struct fuse_file_info fi;
+
+        memset(&fi, 0, sizeof fi);
+        fi.flags = filedesc->OpenFlags;
+        fi.fh = filedesc->FileHandle;
+
+        if (0 != f->ops.flush)
+            f->ops.flush(filedesc->PosixPath, &fi);
+    }
 }

 static VOID fsp_fuse_intf_Close(FSP_FILE_SYSTEM *FileSystem,

```

This way the issue seem to be fixed, application is still blocked on first `flush` and double flushing is much better than no flushing before closing a file. 

I think we should not rely on linux of windows behavior but try to provide (emulate) most consistent fuse layer between operation systems.

I started to making descriptive table to get most clarity about this issue and how it may be fixed/worked around at moment:

| Operation |  Linux | Windows |
| ------------- | ------------- |------------- |
| `fd = open(...)` |  |  |
| `write(fd, ...)`  | `fsp_fuse_intf_Write` | `fsp_fuse_intf_Write` |
| `close(fd)`  | `fsp_fuse_intf_Flush`, `fsp_fuse_intf_Close` | nothing |
| sometime ago | nothing | `fsp_fuse_intf_Flush`, `fsp_fuse_intf_Close` |

**Comment by kalemas** [comment]
And little more to get `Cleanup` status, but it still does not return error code to the application
```diff
diff --git a/inc/winfsp/winfsp.h b/inc/winfsp/winfsp.h
index ff07e0b3..5b5dd3ec 100644
--- a/inc/winfsp/winfsp.h
+++ b/inc/winfsp/winfsp.h
@@ -401,7 +401,7 @@ typedef struct _FSP_FILE_SYSTEM_INTERFACE
      *     CanDelete
      *     SetDelete
      */
-    VOID (*Cleanup)(FSP_FILE_SYSTEM *FileSystem,
+    NTSTATUS (*Cleanup)(FSP_FILE_SYSTEM *FileSystem,
         PVOID FileContext, PWSTR FileName, ULONG Flags);
     /**
      * Close a file.
diff --git a/src/dll/fsop.c b/src/dll/fsop.c
index 429d8c0c..3a912af6 100644
--- a/src/dll/fsop.c
+++ b/src/dll/fsop.c
@@ -978,8 +978,10 @@ FSP_API NTSTATUS FspFileSystemOpOverwrite(FSP_FILE_SYSTEM *FileSystem,
 FSP_API NTSTATUS FspFileSystemOpCleanup(FSP_FILE_SYSTEM *FileSystem,
     FSP_FSCTL_TRANSACT_REQ *Request, FSP_FSCTL_TRANSACT_RSP *Response)
 {
+    NTSTATUS Result = STATUS_SUCCESS;
+
     if (0 != FileSystem->Interface->Cleanup)
-        FileSystem->Interface->Cleanup(FileSystem,
+        Result = FileSystem->Interface->Cleanup(FileSystem,
             (PVOID)ValOfFileContext(Request->Req.Cleanup),
             0 != Request->FileName.Size ? (PWSTR)Request->Buffer : 0,
             (0 != Request->Req.Cleanup.Delete ? FspCleanupDelete : 0) |
@@ -989,7 +991,7 @@ FSP_API NTSTATUS FspFileSystemOpCleanup(FSP_FILE_SYSTEM *FileSystem,
             (0 != Request->Req.Cleanup.SetLastWriteTime ? FspCleanupSetLastWriteTime : 0) |
             (0 != Request->Req.Cleanup.SetChangeTime ? FspCleanupSetChangeTime : 0));
 
-    return STATUS_SUCCESS;
+    return Result;
 }
 
 FSP_API NTSTATUS FspFileSystemOpClose(FSP_FILE_SYSTEM *FileSystem,
diff --git a/src/dll/fuse/fuse_intf.c b/src/dll/fuse/fuse_intf.c
index b13c8541..b1550df9 100644
--- a/src/dll/fuse/fuse_intf.c
+++ b/src/dll/fuse/fuse_intf.c
@@ -1157,11 +1157,12 @@ static NTSTATUS fsp_fuse_intf_Overwrite(FSP_FILE_SYSTEM *FileSystem,
         &Uid, &Gid, &Mode, FileInfo);
 }
 
-static VOID fsp_fuse_intf_Cleanup(FSP_FILE_SYSTEM *FileSystem,
+static NTSTATUS fsp_fuse_intf_Cleanup(FSP_FILE_SYSTEM *FileSystem,
     PVOID FileDesc, PWSTR FileName, ULONG Flags)
 {
     struct fuse *f = FileSystem->UserContext;
     struct fsp_fuse_file_desc *filedesc = FileDesc;
+    NTSTATUS Result = STATUS_SUCCESS;
 
     /*
      * In Windows a DeleteFile/RemoveDirectory is the sequence of the following:
@@ -1191,6 +1192,23 @@ static VOID fsp_fuse_intf_Cleanup(FSP_FILE_SYSTEM *FileSystem,
             if (0 != f->ops.unlink)
                 f->ops.unlink(filedesc->PosixPath);
         }
+    if (Flags & FspCleanupSetLastWriteTime && !filedesc->IsDirectory)
+    {
+        struct fuse_file_info fi;
+
+        memset(&fi, 0, sizeof fi);
+        fi.flags = filedesc->OpenFlags;
+        fi.fh = filedesc->FileHandle;
+        //Result = ERROR_ACCESS_DENIED;
+
+        if (0 != f->ops.flush)
+        {
+            int err = f->ops.flush(filedesc->PosixPath, &fi);
+            Result = fsp_fuse_ntstatus_from_errno(f->env, err);
+            if (err) RaiseException(Result, 0, 0, 0);
+        }
+    }
+    return Result;
 }
 
 static VOID fsp_fuse_intf_Close(FSP_FILE_SYSTEM *FileSystem,
```

How Windows handle error code for asynchronous write?

**Comment by ncw** [comment]
@billziss-gh wrote:
> > What appears to be happening is that rclone is returning an error to cgofuse/WinFSP from the flush call (note the errc=-5 below). But this is never making it back to the application.
> 
> Yes, on Windows the `Flush` error will be ignored.

OK, I understand the behavior now - thank you :-)

Reading your above again, it seems to me that Windows file systems can't return errors on `close()`.

That is unfortunate for rclone. What it is doing is each `write()` call is writing data to a TCP stream and the `close()` call closes the stream. Crucially we get the status from the cloud provider at the `close` so we'd really like to send this back to the application.
 
> > Rclone (in this case without a local disk buffer) can only return errors after it has finished writing (or attempted to write) the file. It can only do that after the last close which under Linux at least translates to the Flush call. Before Flush is called we don't know if any more writes will be coming.
> 
> The problem with `flush` is that even on Linux it is not guaranteed to arrive before a `release`, even when `write` is performed. Consider this scenario (untested, but encapsulates behavior I have seen over the years in Linux and macOS).

Rclone has a heuristic for dealing with this sort of thing. If the file is open multiple times it will ignore the flush until the last one. If it gets to release without having called flush then it will call flush first. This doesn't return any errors to the application though so I guess if it fails it fails silently...

> In general I recommend that FUSE file systems do not rely on `flush`.

What would you recommend rclone does instead? As far as I can see rclone can't return errors in `write` as it can't `close` the stream then since it doesn't know how long it is going to be at that point.

**Comment by billziss-gh** [comment]
>@billziss-gh what do you think about following
>```diff
>diff --git a/src/dll/fuse/fuse_intf.c b/src/dll/fuse/fuse_intf.c
>index b13c8541..630da8ba 100644
>--- a/src/dll/fuse/fuse_intf.c
>+++ b/src/dll/fuse/fuse_intf.c
>@@ -1191,11 +1194,23 @@ static VOID fsp_fuse_intf_Cleanup(FSP_FILE_SYSTEM *FileSystem,
>             if (0 != f->ops.unlink)
>                 f->ops.unlink(filedesc->PosixPath);
>         }
>+    if (Flags & FspCleanupSetLastWriteTime && !filedesc->IsDirectory)
>+    {
>+        struct fuse_file_info fi;
>+
>+        memset(&fi, 0, sizeof fi);
>+        fi.flags = filedesc->OpenFlags;
>+        fi.fh = filedesc->FileHandle;
>+
>+        if (0 != f->ops.flush)
>+            f->ops.flush(filedesc->PosixPath, &fi);
>+    }
> }
>
> static VOID fsp_fuse_intf_Close(FSP_FILE_SYSTEM *FileSystem,
>
>```

This seems sensible and I believe that was what I first did when building the WinFsp-FUSE layer. Eventually I changed it to do the `flush` in `Close`.

While I do not remember all the reasons for doing so I note that the return codes from `Cleanup` and `Close` are both **ignored**. So it is not relevant where you place the `flush` if you are doing so in the hopes to get its return code.

> And little more to get `Cleanup` status, but it still does not return error code to the application

Yes, the return code from `Cleanup` is ignored. This is because the status code for `IRP_MJ_CLEANUP` IRP's is ignored. (IRP's or I/O Request Packets is how a Windows driver acts on requests from the operating system.)

I cannot find a direct reference in Microsoft's documentation, but there is a very good explanation of how the whole thing works here: https://community.osr.com/discussion/comment/68399/#Comment_68399

> CloseHandle() cannot fail unless the handle is invalid i.e. the only
time CloseHandle() will return a failure status is if the handle that is
supplied is not an open handle for the process. As for drivers - they
are expected to succeed CLEANUP & CLOSE always.
Ob will destroy the handle table entry regardless of how the driver
handles cleanup - think of CloseHandle() as a destructor, or a memory
free operation. It doesn't fail. Note that CloseHandle() does not call
CLEANUP and return the status of that: it simply decrements the handle
count. If the handle count goes to zero (there could be duplicates of
this handle, each of which bumps up the handle count), an IRP_MJ_CLEANUP
is generated - this is a side effect of the decrement handle count
operation.
>
> Finally when the pointer count goes to zero, an IRP_MJ_CLOSE is
generated, but Ob destroys the object. So regardless of how the client
of Ob (I/O in this case) deals with their object close (i.e. handle
count goes to zero) or object destroy (pointer count goes to zero), Ob
will always destroy handles when the handle count is zero, and delete
objects when pointer count is zero.

Here is another reference that `IRP_MJ_CLEANUP` cannot be failed: https://community.osr.com/discussion/comment/16472/#Comment_16472

> this is not so good, because in the time of cleanup
the user has no way to know about some error, because
IRP_MJ_CLEANUP is the result of the last NtClose on some
file object, thus calling ObpDecrementHandleCount which
then calls file object type close method IopCloseFile,
but the last routine doesn't return any value back to
the object manager, so your error is really "eaten" by
the I/O manager and doesn't reach the caller of NtClose.

**Comment by billziss-gh** [comment]
@ncw

> Reading your above again, it seems to me that Windows file systems can't return errors on close().

Yes. Please see my response above to @kalemas. `CloseHandle` cannot return error codes from the driver doing the actual closing (see my discussion above about `IRP_MJ_CLEANUP` and rationale).

>> In general I recommend that FUSE file systems do not rely on flush.
>
>What would you recommend rclone does instead? As far as I can see rclone can't return errors in write as it can't close the stream then since it doesn't know how long it is going to be at that point.

This is a major headache when designing file systems but is unavoidable. As you noticed yourself the `flush` is not guaranteed to come even on Linux. Also flushing on file close can kill cache performance in some usage scenarios (e.g. continuous loop of open, append, close from a shell).

The answer depends on the file system and its requirements. A common pattern that I use is the following: I design my file system so that it persists file data in a local cache, usually the local disk. Then I have threads that wake up after the file is closed and inactive for some time that flush the file to the remote. If the network fails I keep retrying with some back-off algorithm. If the file system process fails (e.g. kill or crash) I continue from when I left off in my persistent cache when the file system restarts.

This works for many kinds of file systems and it may work for rclone. For example, I use this technique in [objfs](https://github.com/billziss-gh/objfs).

**Comment by kalemas** [comment]
> This works for many kinds of file systems and it may work for rclone
rclone already does somewhat like that, but it does not guarantee data transfer. So when we started to implement webdav projected filesystem with read-only files that are implemented in webdav by returning error for operations, we encountered this issue...

I think if there is difference between sync and async write in fuse then rclone should perform sync write on remote for requested sync write. 

In windows this is implemented this way https://www.microsoftpressstore.com/articles/article.aspx?p=2224047&seqNum=5 (and previous chapter also should be helpful, i did not read the whole article yet) and should be supported by WinFSP.

If there is no way to get difference on operation level i think user should specify operations mode to rclone.

**Comment by ncw** [comment]
> The answer depends on the file system and its requirements. A common pattern that I use is the following: I design my file system so that it persists file data in a persistent local cache, usually the local disk. Then I have threads that wake up after the file is closed and inactive for some time that flush the file to the remote. If the network fails I keep retrying with some back-off algorithm. If the file system process fails (e.g. kill or crash) I continue from where I left off in my persistent local cache when the file system restarts.

Sorry for the delay in responding! This is exactly how rclone works with `--vfs-cache-mode writes` or `--vfs-cache-more full` so that is probably a workaround for you @kalemas

**Comment by kalemas** [comment]
@ncw that does not guarantee data transfer so writing readonly webdav file will never succeed, so this makes rclone non robust solution for projecting remote filesystem in rw mode. We use both types of acces rw and ro, for readonly rclone is just fine. But are not so much alternatives for on demand data fetch, lanman/webclient have no long live cache, ProjFS looks interesting but there is a lot to implement against rclone.

I m not sure but sudden rclone failure also may not restore upload after next start. So there would be locally cached file that was not uploaded on remote.

**Comment by ncw** [comment]
I can't think of a way to fix your scenario @kalemas . Either we use no caching in which case Windows defeats us telling the user about the problem, or we use caching in which case the file stays forever in the cache. Can you think of a fix?

**Comment by kalemas** [comment]
> I can't think of a way to fix your scenario @kalemas .

With readonly file i see the only option to actually write data to destination and then update (or not) the cache. I realize that this would be not perfect in sense of performance and not suitable for rclone architecture. 

Maybe it is possible to rely on file attributes, but afaik webdav does not have readonly file property but just restrict write operations for readonly files.

**Comment by billziss-gh** [comment]
I am closing this as I do not believe that there is a sensible way to fix this problem in Windows.

Even if we implemented @kalemas suggestion of moving `flush` to `Cleanup` it would still not be good, because you cannot report an error from `Cleanup` all the way to `CloseHandle`. (It would also impede an optimization where I am trying to eliminate all uses of `Cleanup` by the WinFsp-FUSE layer in the near future.)
